Client/Frontend/Widgets/BottomSheetViewController.swift

```diff
@@ -0,0 +1,257 @@
+            if showAsPopup {
+                update()
+                return
+            }
+            
+            let initialY = initialDrawerYPosition
+            
+            // Only move the view if dragged below initial level.
+            if yPosition <= initialY {
+                yPosition = initialY
+            } else if yPosition > maxY { // Dragged all way down, remove the view.
+                yPosition = maxY
+            }
+            
+            update()
+        }
+    }
+    
+    private var initialDrawerYPosition: CGFloat {
+        let popupY = ceil((view.frame.height / 2) - (contentView.frame.height / 2))
+        
+        let regularY = view.frame.maxY - contentView.frame.height
+        
+        return showAsPopup ? popupY : regularY
+    }
+    
+    private var showAsPopup: Bool {
+        traitCollection.userInterfaceIdiom == .pad && traitCollection.horizontalSizeClass == .regular
+    }
+    
+    // MARK: - Lifecycle
+
+    init() {
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    @available(*, unavailable)
+    required init?(coder aDecoder: NSCoder) { fatalError() }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        view.addSubview(backgroundOverlayView)
+        view.addSubview(contentView)
+        
+        contentView.backgroundColor = .white
+
+        let panGestureRecognizer = UIPanGestureRecognizer(target: self, action: #selector(handlePanGesture))
+        contentView.addGestureRecognizer(panGestureRecognizer)
+
+        contentView.addSubview(handleView)
+        
+        contentView.addSubview(closeButton)
+        
+        closeButton.addTarget(self, action: #selector(closeView), for: .touchUpInside)
+        contentView.isHidden = true
+        
+        makeConstraints()
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        show()
+    }
+    
+    override func viewDidLayoutSubviews() {
+        yPosition = contentView.isHidden ? view.frame.maxY : initialDrawerYPosition
+    }
+    
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+        view.setNeedsUpdateConstraints()
+    }
+    
+    override func updateViewConstraints() {
+        let allCorners: CACornerMask = [.layerMinXMinYCorner, .layerMaxXMinYCorner,
+                          .layerMinXMaxYCorner, .layerMaxXMaxYCorner]
+        
+        let onlyTopCorners: CACornerMask = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
+        contentView.layer.maskedCorners = showAsPopup ? allCorners : onlyTopCorners
+        
+        handleView.isHidden = showAsPopup
+        
+        // Don't remake constraints if not needed.
+        if yPosition == initialDrawerYPosition { return }
+        
+        contentView.snp.remakeConstraints {
+            if showAsPopup || (traitCollection.userInterfaceIdiom == .phone
```

Comment:
Maybe create const for `traitCollection.userInterfaceIdiom == .phone && UIApplication.shared.statusBarOrientation.isLandscape`

```
let iPhoneLandscape = traitCollection.userInterfaceIdiom == .phone && UIApplication.shared.statusBarOrientation.isLandscape
if showAsPopup && iPhoneLandscape { }
```

Probably not needed in this simple situation though.